src/cli/cli_instance.cpp

```diff
@@ -0,0 +1,126 @@
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the CLI interpreter Instance related functions.
+ */
+
+#ifdef OPENTHREAD_CONFIG_FILE
+#include OPENTHREAD_CONFIG_FILE
+#else
+#include <openthread-config.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include <openthread.h>
+#include "cli.hpp"
+
+namespace Thread {
+
+namespace Cli {
+
+void Interpreter::CacheInstances()
+{
+    if (mApiInstance)
+    {
+        otDeviceList *aDeviceList = otEnumerateDevices(mApiInstance);
+        assert(aDeviceList);
+
+        mInstancesLength = aDeviceList->aDevicesLength > MAX_CLI_OT_INSTANCES ?
+                            MAX_CLI_OT_INSTANCES :
+                            (uint8_t)aDeviceList->aDevicesLength;
+
+        for (uint8_t i = 0; i < mInstancesLength; i++)
+        {
+            mInstances[i].aInterpreter = this;
+            mInstances[i].aInstance = otInstanceInit(mApiInstance, &aDeviceList->aDevices[i]);
+            assert(mInstances[i].aInstance);
+            otSetStateChangedCallback(mInstances[i].aInstance, &Interpreter::s_HandleNetifStateChanged, &mInstances[i]);
+        }
+
+        otFreeMemory(aDeviceList);
+
+        if (mInstancesLength > 0) { mInstance = mInstances[0].aInstance; }
+    }
+}
+
+#define GUID_FORMAT "{%08lX-%04hX-%04hX-%02hhX%02hhX-%02hhX%02hhX%02hhX%02hhX%02hhX%02hhX}"
+#define GUID_ARG(guid) guid.Data1, guid.Data2, guid.Data3, guid.Data4[0], guid.Data4[1], guid.Data4[2], guid.Data4[3], guid.Data4[4], guid.Data4[5], guid.Data4[6], guid.Data4[7]
+
+void Interpreter::ProcessInstanceList(int argc, char *argv[])
+{
+    sServer->OutputFormat("%d instances found:\r\n", mInstancesLength);
+
+    for (uint8_t i = 0; i < mInstancesLength; i++)
+    {
+        GUID aDeviceGuid = otGetDeviceGuid(mInstances[i].aInstance);
```

Comment:
Would it be possible to move these API definitions to be within a windows-specific header file?  For me, it's a little strange to have these functions declared in `include/openthread.h` but not implemented anywhere in `src/core`.  Again, I won't block this PR since this issue is unrelated from the purpose of the PR.  We can address this at a later time.